Develop

IvchenkoElena opened the pull request.

IvchenkoElena committed:
Добавлены эндпоинты и логика удаления пользователей и фильмов

rovniipoc committed:
Добавить дата-класс Review

rovniipoc committed:
Добавить интерфейс ReviewStorage

rovniipoc committed:
Добавить роумаппер для отзывов ReviewRowMapper

rovniipoc committed:
Добавить ReviewDbStorage, изменить ReviewRowMapper, изменить Review и сопутствующие изменения

Изменения Review:
- тип данных reviewId, userId, filmId изменен с Long на Integer;

Изменения ReviewRowMapper:
- добавлен статический метод makeReview;

rovniipoc committed:
Добавить ReviewService

rovniipoc committed:
Добавить ReviewController, изменить ReviewDbStorage

Изменения ReviewDbStorage:
- удалено присвоение значения count по умолчанию, т.к. значение по умолчанию присваивается в ReviewController;

rovniipoc committed:
Изменить data.sql, изменить schema.sql

Изменения data.sql:
- добавлена очистка таблицы reviews при старте;

Изменения schema.sql:
- добавлена таблица reviews;

rovniipoc committed:
Добавить логирование в ReviewController

rovniipoc committed:
Добавить логирование в ReviewController, изменить ReviewDbStorage

Изменения ReviewDbStorage:
- исправить метод getByParams();

rovniipoc committed:
Изменить ReviewService, изменить ReviewDbService, изменить ReviewDbStorage

Изменения ReviewService:
- добавлен методы checkUserExistById(), checkCorrectReviewParams();
- в метод checkCorrectReviewParams() добавлены проверки на пустой контент и пустой тип отзыва;

Изменения ReviewDbService:
- теперь нельзя обновить userId и filmId у отзыва (на основании тестов postman);

Изменения ReviewDbStorage:
- исправлен метод update (теперь он возвращает взятый из репозитория обновленный отзыв);

rovniipoc committed:
Реализовать функционал отзывов, изменить ReviewDbStorage, изменить Review

Изменения ReviewDbStorage:
- теперь через метод update() нельзя изменить кол-во лайков (на основании тестов postman);

Изменения Review:
- Добавлено значение по умолчанию для поля useful;

rovniipoc committed:
Добавить плагин checkstyle

rovniipoc committed:
Исправить по чекстайлу

IvchenkoElena committed:
Поправлены методы для прохождения тестов постмана

rovniipoc committed:
Изменить логику лайков отзывов для выполнения тестов postman ветки add-reviews

Теперь если пользователь поставил лайк, затем поставил дизлайк, то лайк убирается автоматически (и наоборот). Один пользователь может поставить не больше одного лайка(дизлайка)

Изменения ReviewDbStorage:
- изменены методы по работе с лайками, добавлен новый приватный метод обновления поля useful на основании лайков.

Изменения data.sql, schema.sql:
- добавлена таблица REVIEW_LIKES_DISLIKES содержащая информацию о том кто какому отзыву поставил лайк или дизлайк.

IvchenkoElena committed:
Merge pull request #5 from IvchenkoElena/add-remove-endpoint

Добавлены эндпоинты и логика удаления пользователей и фильмов

rovniipoc committed:
Исправить schema.sql

Изменения schema.sql:
- для референсных ключей таблиц reviews и review_likes_dislikes добавлены команды "on delete cascade"

IvchenkoElena committed:
Реализована сущность Режиссер контроллер Режиссеров и связанные сним методы в сервисе и Storage

rovniipoc committed:
Добавить функциональность "Вывод самых популярных фильмов по жанру и годам" (доработан имеющийся метод bestFilms())

IvchenkoElena committed:
Добавлен эндпоинт поиска фильма по режиссеру в FilmController, изменена логика хранения режиссеров, добавлена таблица связи.

IvchenkoElena committed:
Удален неиспользуемый импорт

IvchenkoElena committed:
Удален ненужные методы

IvchenkoElena committed:
Удален неиспользуемый импорт

IvchenkoElena committed:
Чек стайл

IvchenkoElena committed:
Чек стайл

IvchenkoElena committed:
Убрала вложенный запрос

rovniipoc committed:
Заменить в FilmDbStorage ветвление в методе bestFilms() на универсальный запрос

rovniipoc committed:
Изменить в FilmDbStorage запрос в методе bestFilms()

Теперь параметры для запроса не дублируются

IvchenkoElena committed:
Merge pull request #8 from IvchenkoElena/add-most-populars

add-most-populars

rovniipoc committed:
Изменить хранение лайков отзывов

Изменения:
- в schema.sql в таблице review_likes_dislikes удален столбец is_dislike, теперь информация и о лайках и о дизлайках хранится в столбце is_like, значение может быть: TRUE = лайк, FALSE = дизлайк, NULL = отсутствие лайка/дизлайка
- в ReviewDbStorage внесены соответствующие изменения в методы работы с лайками/дизлайками к отзывам

IvchenkoElena committed:
Метод loadDirectors реализован через NamedParameterJdbcTemplate

artyom-petukhov-yapr committed:
Реализация "Функциональность «Общие фильмы». 1 SP"

IvchenkoElena committed:
Merge pull request #7 from IvchenkoElena/add-reviews

add-reviews

IvchenkoElena committed:
Вынесены методы saveGenres, saveDirectors, saveLikes

IvchenkoElena committed:
Merge branch 'refs/heads/develop' into add-director

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java

IvchenkoElena committed:
Правки для мерджа

IvchenkoElena committed:
Merge pull request #9 from IvchenkoElena/add-director

Add director

artyom-petukhov-yapr committed:
Merge branch 'develop' into add-common-films

artyom-petukhov-yapr committed:
поддержка изменений в develop

artyom-petukhov-yapr committed:
Реализация "Функциональность «Лента событий». 3 SP"

IvchenkoElena committed:
Merge pull request #11 from IvchenkoElena/add-feed

Реализация "Функциональность «Лента событий». 3 SP"

artyom-petukhov-yapr committed:
Merge branch 'develop' into add-common-films

artyom-petukhov-yapr committed:
FilmDbStorage.getCommonFilms(): упрощение запроса за счет использования колонки films.likes_count

artyom-petukhov-yapr committed:
Merge pull request #10 from IvchenkoElena/add-common-films

Реализация "Функциональность «Общие фильмы». 1 SP"

Alex-Ivash committed:
Реализовал простой поиск фильмов по названию и/или режиссеру.
https://github.com/users/IvchenkoElena/projects/2/views/1?pane=issue&itemId=85840810

Alex-Ivash committed:
Реализация простой рекомендательной системы фильмов

Alex-Ivash committed:
Сделал поиск единым методом в FilmDbStorage

IvchenkoElena committed:
Merge pull request #12 from IvchenkoElena/add-search

Реализация "Поиска"

IvchenkoElena committed:
Merge pull request #13 from IvchenkoElena/add-recommendations

Реализация "Рекомендаций"

IvchenkoElena committed:
минорные правки

IvchenkoElena committed:
минорные правки

IvchenkoElena committed:
Отключены тесты

IvchenkoElena committed:
Добавлены валидации

IvchenkoElena committed:
Правки для тестов feed

IvchenkoElena committed:
Правки для тестов

rovniipoc committed:
Доработать механику лайков фильмов

IvchenkoElena committed:
Доработана механика лайков фильмов

IvchenkoElena committed:
Добавлен метод loadLikes

IvchenkoElena committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/model/Film.java
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java

IvchenkoElena committed:
Удален лишний exception

IvchenkoElena committed:
методы loadGenres loadLikes сделаны параметризованными

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Director.java:
```diff
@@ -0,0 +1,11 @@
+@Data
+public class Director {
+    private Integer id;
+    @NotNull
```
Для валидации строк используйте `@NotBlank`, она проверяет не только на null, но и на пустую строку `""` и строку, состоящую из пробелов `" "`.

Для того, чтобы валидация проходила автоматически, нужно добавить аннотации `@Valid` у параметра в методе контроллера, иначе в аннотации смысле нет

IvchenkoElena replied:
Поправила

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -55,7 +61,32 @@ public void removeLike(@PathVariable Integer id,
+
+    @GetMapping("/director/{directorId}")
+    public List<Film> findFilmsByDirector(@PathVariable Integer directorId,
+                                      @RequestParam String sortBy) {
```
Рекомендую сделать для сортировки енам. Всегда можно посмотреть возможные варианты сортировки и при появлении новых, просто добавить константу и условие или case

IvchenkoElena replied:
енам сделала. В контроллере получаю строку, в сервисе валидирую и превращаю в енам. В сторадж передаю уже енам

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/FilmController.java:
```diff
@@ -55,7 +61,32 @@ public void removeLike(@PathVariable Integer id,
+
+    @GetMapping("/search")
+    public List<Film> searchFilms(@RequestParam String query,
+                                  @RequestParam List<String> by) {
```
По требованиям вроде этот параметр должен передаваться строкой, потом его можно сплитануть по разделителю в список

Alex-Ivash replied:
Он строкой передается. Spring его сам разбирает в список

Ivanov-s-tmn replied:
Окей, не встречал такую реализацию

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/controller/ReviewController.java:
```diff
@@ -0,0 +1,72 @@
+
+    @PostMapping
+    public Review addNew(@RequestBody Review newReview) {
+        log.info("Поступил запрос на добавление нового отзыва: {}", newReview.toString());
```
вызов `toString()` не нужен, он и так будет вызван при печати в лог

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/DirectorDbStorage.java:
```diff
@@ -0,0 +1,87 @@
+        }
+    }
+
+    public void loadDirectors(List<Film> films) {
+        final Map<Integer, Film> filmById = films.stream().collect(Collectors.toMap(Film::getId, film -> film));
+        List<Integer> filmIds = films.stream().map(Film::getId).toList();
+        SqlParameterSource parameters = new MapSqlParameterSource("filmIds", filmIds);
+
+        namedJdbcTemplate.query(FIND_FILM_DIRECTORS_QUERY + "(:filmIds)", parameters, (rs) -> {
+            final Film film = filmById.get(rs.getInt("FILM_ID"));
+            film.addDirector(mapper.mapRow(rs, 0));
+        });
+    }
```
Смотрите для режиссеров и фильмов лучше не делать аналогичную связь как с жанрами через таблицу связей. В случае жанрами данные статичные и новые жанры не появляются или появляются очень редко. А в случае с режиссерами совсем не так, поэтому будет достаточно добавить айди режиссера в таблицу фильмов и уже через обычный join подтягивать их вместе с фильмом по films.director_id = directors.director_id, как и с рейтингом, только там 1 к 1, а тут 1 ко многим

Ivanov-s-tmn replied:
upd: тут проблематично будет реализовать как я описал, потому что у фильма может быть несколько режиссеров, не туда посмотрел в требования, поэтому оставьте как есть)

IvchenkoElena replied:
да, я сначала так и сделала одну новую колонку с айди режиссера в таблицу с фильмами, но потом тесты показали, что режиссеров может быть несколько, и пришлось городить таблицу связей

IvchenkoElena replied:
Решение: оставляем так

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java:
```diff
@@ -83,27 +107,202 @@ public Film save(@RequestBody Film newFilm) {
+        return likesCount;
+    }
+
+    @Override
+    public void loadLikes(List<Film> films) {
+        final Map<Integer, Film> filmById = films.stream().collect(Collectors.toMap(Film::getId, film -> film));
+        List<Integer> filmIds = films.stream().map(Film::getId).toList();
+        SqlParameterSource parameters = new MapSqlParameterSource("filmIds", filmIds);
+
+        namedJdbcTemplate.query(FIND_FILM_LIKES_QUERY + "(:filmIds)", parameters, (rs) -> {
+            final Film film = filmById.get(rs.getInt("FILM_ID"));
+            film.addLike(rs.getInt("USER_ID"));
+        });
```
В случае с лайками аналогично как и с режиссерами, у вас есть поле likes_count, вы можете его тянуть вместе с фильмом простым селектом, вам только нужно его обновлять при добавлении/удалении лайка
Плюсом, запрос с сортировкой станет проще и в целом все запросы, где нужны лайки

IvchenkoElena replied:
Поле likes_count я добавила еще после прошлого спринта. Но полностью удалять список лайков не стала, не до конца понимаю последствия такого изменения. Это получается и таблицу likes тоже нужно будет полностью удалить? Нам же в тестах приходят фильмы со списком лайков. А мы это никуда не будем сохранять? Сортировку мы вроде бы по полю likes_count везде реализовывали. Точно можно удалить и не бояться? Тесты не повалятся?

IvchenkoElena replied:
Там тесты финальные заточены так, что при попытке одним и тем же пользователем поставить снова лайк тому же фильму, лайк новый добавляться не будет, но действие будет фиксироваться для event. Если у нас не будет списка поставленных лайков и таблицы likes в БД, то как это можно реализовать?

Ivanov-s-tmn replied:
Нет, почему, таблицу likes удалять не надо, вы же туда добавляете записи с userId и filmId при добавлении и удаляете при удалении. Просто для облегчения сортировок и получения количества лайков вы это тянете из таблицы films по полю likes_count. Просто нужно тогда обновлять поле likes_count при удалении/добавлении лайков (селектом количества записей из таблицы likes по film_id или просто инкремент/декремент)

IvchenkoElena replied:
То есть удаляем только поле списка лайков в Model.Film? и связанные с ним методы?

IvchenkoElena replied:
Удалила

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/model/Film.java:
```diff
@@ -24,14 +25,23 @@ public class Film {
     }
-}
+
+    public void addDirector(Director director) {
+        getDirectors().add(director);
+    }
+
+    public void addLike(Integer userId) {
+        getLikes().add(userId);
+    }
```
Эти методы будут не нужны, если реализуете получение данных из БД как я описал в комментариях

IvchenkoElena replied:
addDirector значит оставляем, а addLike будем ли удалять, зависит от решения по предыдущему комментарию, точно нужно удалить список лайков и таблицу в likes в БД?

Ivanov-s-tmn replied:
Этот метод в любом случае не нужен, так как можно тянуть лайки из likes_count и не нужно делать отдельный запрос

IvchenkoElena replied:
Удалила

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/ReviewService.java:
```diff
@@ -0,0 +1,112 @@
+        reviewStorage.deleteDislike(reviewId, userId);
+    }
+
+    void checkUserExistById(Integer userId) {
+        if (userId <= 0) {
+            throw new NotFoundException(String.format("Пользователь c ID %d не найден", userId));
+        }
+        userStorage.getById(userId);
+    }
+
+    void checkFilmExistById(Integer filmId) {
+        if (filmId <= 0) {
+            throw new NotFoundException(String.format("Фильм c ID %d не найден", filmId));
+        }
+        filmStorage.getById(filmId);
+    }
+
+    void checkCorrectReviewParams(Review review) {
+        Integer userId = review.getUserId();
+        if (userId <= 0) {
+            throw new NotFoundException(String.format("Пользователь c ID %d не найден", userId));
+        }
+        userStorage.getById(userId);
+
+        Integer filmId = review.getFilmId();
+        if (filmId <= 0) {
+            throw new NotFoundException(String.format("Фильм c ID %d не найден", filmId));
+        }
+        filmStorage.getById(filmId);
+
+        String content = review.getContent();
+        if (content == null || content.isBlank()) {
+            throw new ValidationException("Невозможно добавить пустой отзыв");
+        }
+
+        Boolean isPositive = review.getIsPositive();
+        if (isPositive == null) {
+            throw new ValidationException("Невозможно добавить отзыв без указания типа отзыва");
+        }
```
Вы можете проверять входящие параметры еще на стороне контроллера вместо ручных проверок <= 0, просто добавьте аннотации `@Positive` у параметра и `@Validated` над контроллером

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/service/UserService.java:
```diff
@@ -14,9 +15,21 @@
     private final UserStorage userStorage;
+    private final EventStorage eventStorage;
+    private final FilmStorage filmStorage;
+    private final GenreDbStorage genreStorage;
```
Внедряйте зависимости через интерфейсы, чтобы не привязываться к конкретной реализации

IvchenkoElena replied:
Поправила

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/DirectorRowMapper.java:
```diff
@@ -0,0 +1,19 @@
+import java.sql.ResultSet;
+import java.sql.SQLException;
+
+@Component
+public class DirectorRowMapper implements RowMapper<Director> {
+    @Override
+    public Director mapRow(ResultSet rs, int rowNum) throws SQLException {
+        Director director = new Director();
+        director.setId(rs.getInt("director_id"));
+        director.setName(rs.getString("director_name"));
+        return director;
+    }
+}
```
Вы можете реализовать метод-маппер в самих стораджах, чтобы не плодить классы и компоненты контекста, как пример для юзера:
```java
static User makeUser(ResultSet rs, int rowNum) throws SQLException {
		return new User(rs.getLong("user_id"), rs.getString("email"), rs.getString("login"), rs.getString("name"), toLocalDate(rs.getDate("birthday")));
}
```

IvchenkoElena replied:
Стоит это сделать для всех сущностей? Убрать все классы RowMapper?

IvchenkoElena replied:
Точно ли стоит это делать? Текущая реализация проще в поддержке решения, т.к. в случае изменения схемы бд проще скорректировать соответствующие мапперы, чем искать методы в сторейджах, где реализуется маппинг. + маппер берет на себя чётко определённую ответственность и ничего лишнего, в то время как вынос его кода в сторейдж размывает сферу ответственности самого сторейджа

Ivanov-s-tmn replied:
Мне больше нравиться вариант с методом в сторадж классе, это в принципе не расширяет его область ответственности, класс предназначен для работы с сущностями, а этот метод маппит выдачу из бд в сущность. К тому же, у каждого стораджа будет только один метод преобразования в ту сущность, с которой он и работает, а благодаря статику и package-private модификатору доступа вы сможете их вызывать при необходимости в других стораджах без препятствий

Ivanov-s-tmn replied:
Но настаивать не буду, можете оставить и так, это как рекомендация)

IvchenkoElena replied:
Решение команды: оставить отдельные классы.

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/EventDbStorage.java:
```diff
@@ -0,0 +1,31 @@
+import java.util.List;
+
+@Repository("eventDbStorage")
+@Primary
```
Вроде бы вы используете `@Qualifier` для задания приоритета компонентам, достаточно его, дополнительных аннотаций не нужно

IvchenkoElena replied:
Убрала все аннотации Primary

IvchenkoElena replied:
Насколько я понимаю, нам и Qualifier особо не нужен, раз других реализаций Стораджа нет

Ivanov-s-tmn replied:
Да, `@Qualifier` не нужен в данном случае, просто вам про него рассказали, чтобы имели в виду на будущее и попробовали применить на практике)

Ivanov-s-tmn commented on src/main/java/ru/yandex/practicum/filmorate/storage/ReviewDbStorage.java:
```diff
@@ -0,0 +1,126 @@
+        }
+    }
+
+    @Override
+    public void addLike(Integer reviewId, Integer userId) {
+        Boolean isLike = true;
+        String query = "MERGE INTO REVIEW_LIKES_DISLIKES(review_id, user_id, is_like) KEY(REVIEW_ID,USER_ID) VALUES(?, ?, ?)";
+        update(query, reviewId, userId, isLike);
+        updateUsefulRateByReviewId(reviewId);
+    }
+
+    @Override
+    public void addDislike(Integer reviewId, Integer userId) {
+        Boolean isLike = false;
+        String query = "MERGE INTO REVIEW_LIKES_DISLIKES(review_id, user_id, is_like) KEY(REVIEW_ID,USER_ID) VALUES(?, ?, ?)";
+        update(query, reviewId, userId, isLike);
+        updateUsefulRateByReviewId(reviewId);
+    }
+
+    @Override
+    public void deleteLike(Integer reviewId, Integer userId) {
+        String query = """
+                UPDATE REVIEW_LIKES_DISLIKES
+                SET IS_LIKE = NULL
+                WHERE REVIEW_ID = ? AND USER_ID = ? AND IS_LIKE = TRUE
+                """;
+        update(query, reviewId, userId);
+        updateUsefulRateByReviewId(reviewId);
+    }
+
+    @Override
+    public void deleteDislike(Integer reviewId, Integer userId) {
+        String query = """
+                UPDATE REVIEW_LIKES_DISLIKES
+                SET IS_LIKE = NULL
+                WHERE REVIEW_ID = ? AND USER_ID = ? AND IS_LIKE = FALSE
+                """;
+        update(query, reviewId, userId);
+        updateUsefulRateByReviewId(reviewId);
+    }
```
Можно объединить в один метод лайк/дизлайк и в один метод удаление лайка/дизлайка

rovniipoc replied:
Добрый день.

Если объединить методы лайк/дизлайк в один, то тогда необходимо в контроллере объявлять переменную boolean isLike и передавать ее в метод как дополнительный аргумент метода.

Если объединить методы удаления лайков/дизлайков то ситуация аналогичная, т.к. удаление должно происходить только если лайк(дизлайк) существует.

Если же сделать так:
чтобы не передавать параметр isLike в метод - сделать проверку в самом методе на наличие лайка/дизлайка и далее через if добавлять/удалять лайк, тогда метод будет работать как "тумблер":
1) лайка нет
2) пришел запрос put films/1/like/1
3) лайк появился
4) пришел повторный запрос put films/1/like/1
5) лайк исчез

Вроде тесты ожидают другого поведения:
1) лайка нет
2) пришел запрос put films/1/like/1
3) лайк появился
4) пришел повторный запрос put films/1/like/1
5) ничего не изменилось (лайк остался)

Прошу пояснить:
1) Целесообразна ли такая оптимизация: введение в метод дополнительного аргумента или добавление нового кода ветвления if в зависимости от наличия/отсутствия лайка/дизлайка вместо использования двух отдельно имеющихся небольших и легкочитаемых методов?
2) Если изменить метод добавления лайков, чтобы он работал как "тумблер" (put like будет и добавлять и удалять лайк), то что делать с валящимися тестами?

Ivanov-s-tmn replied:
Добрый день! При удалении же нам не важно, удаляем мы лайк или дизлайк, достаточно удалить запись по нужным полям `delete from REVIEW_LIKES_DISLIKES where REVIEW_ID = ? and USER_ID = ?`, будет она там или нет уже не важно.

По добавлению лайка/дизлайка, всегда лучше добавить дополнительный параметр, вместо дублирования нескольких строк кода, в данном случае вам это ничего не стоит, вы просто будет передавать в метод reviewId, userId и дополнительно булевый флаг, который будет служить индикатором лайка/дизлайка

rovniipoc replied:
> При удалении же нам не важно, удаляем мы лайк или дизлайк, достаточно удалить запись по нужным полям delete from REVIEW_LIKES_DISLIKES where REVIEW_ID = ? and USER_ID = ?, будет она там или нет уже не важно.

Как это не важно?
У нас разные api на удаление лайков и дизлайков.
И это указано в ТЗ на задачу:
![image](https://github.com/user-attachments/assets/6b38355f-0ca6-4505-9571-23aaabd5ab80)

Пример:
Есть лайк от пользователя №1 на отзыве №1.
Если по какой то причине (ошибке) клиент присылает запрос del reviews/1/dislike/1 то лайк удаляется, что не соответствует замыслу конкретного api (удаление дизлайка).

Ivanov-s-tmn reviewed (commented):
Всем привет! Извините за долгое ревью.

В целом все отлично, вы молодцы, что следуете общему стиля проекта и стараетесь от него не отступать. При работе над одним проектов старайтесь всегда придерживаться общего стиля, чтобы проект со временем не превратился в коктейль из разных стилей разных разработчиков :)

Рассмотрите возможность изменения структуры БД относительно хранения режиссеров, я описал это в одном из своих комментариев и предложил альтернативный вариант.

Старайтесь оставлять логи в основных частях проекта, контроллеры - сервисы - стораджи (репозитории), это позволит лучше отслеживать работу и проще отслеживать на каком этапе возникает проблема.

Если будут вопросы по моим комментариям, обращайтесь)

Всем отличного дня!

IvchenkoElena committed:
Валидации строковых полей в сущностях заменены на NotBlank, в параметрах методов контроллеров добавлена анннотация Valid

IvchenkoElena committed:
Добавлен enum для сортировки списка фильмов режиссера

IvchenkoElena committed:
Поправлена зависимость в UserService на интерфейс

IvchenkoElena committed:
Убрала все лишние аннотации Primary

rovniipoc committed:
Merge remote-tracking branch 'origin/develop' into develop

# Conflicts:
#	src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java
#	src/main/java/ru/yandex/practicum/filmorate/storage/FilmDbStorage.java

IvchenkoElena committed:
Удалено поле списка лайков в Model.Film и все связанные с ним методы

rovniipoc committed:
Merge remote-tracking branch 'origin/develop' into develop

rovniipoc committed:
Изменить по замечаниям ревью

Изменения:
- Часть валидации выполняется через аннотации;
- Теперь в ReviewDbStorage используется один универсальный метод setLikeOrDislike, который принимает boolean isLike;
- Теперь оба api (del like и del dislike) удаляют и лайки и дизлайки (выполняют одинаковую функцию);
- Прочие мелкие правки;

Ivanov-s-tmn reviewed (commented):
Все отлично, проект зачтен, всем удачи дальше и хороших выходных!
